govcd/catalog.go

```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
+			if err != nil {
+				return nil, err
+			}
+
+			return uploadURL, nil
+		}
+	}
+	return nil, errors.New("catalog upload url isn't found")
+}
+
+func uploadFile(client *Client, uploadLink, filePath string, offset, fileSizeToUpload int64) (int64, error) {
+	log.Printf("[TRACE] Starting uploading: %s, offset: %v, fileze: %v, toLink: %s \n", filePath, offset, fileSizeToUpload, uploadLink)
+
+	file, err := os.Open(filePath)
+	if err != nil {
+		return 0, err
+	}
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return 0, err
+	}
+
+	defer file.Close()
+
+	request, err := newFileUploadRequest(uploadLink, file, offset, fileInfo.Size(), fileSizeToUpload)
+	if err != nil {
+		return 0, err
+	}
+
+	response, err := client.Http.Do(request)
+	if err != nil {
+		return 0, err
+	}
+	defer response.Body.Close()
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return 0, err
+	}
+	log.Printf("[TRACE] Response: %#v\n", response)
+	log.Printf("[TRACE] Response body: %s\n", string(body[:]))
+
+	if response.StatusCode != http.StatusOK {
+		return 0, errors.New("File " + filePath + " upload failed. Err: " + fmt.Sprintf("%#v", response) + " :: " + string(body[:]) + "\n")
+	}
+	return fileInfo.Size(), nil
+}
+
+func findFilePath(filesAbsPaths []string, fileName string) string {
+	for _, item := range filesAbsPaths {
+		_, file := filepath.Split(item)
+		if file == fileName {
+			return item
+		}
+	}
+	return ""
+}
+
+/** Initiates creation of item and returns ovf upload url for created item. */
+func createItemForUpload(client *Client, createHREF *url.URL, catalogItemName string, itemDescription string) (*url.URL, error) {
+
+	reqBody := bytes.NewBufferString(
+		"<UploadVAppTemplateParams xmlns=\"http://www.vmware.com/vcloud/v1.5\" name=\"" + catalogItemName + "\" >" +
+			"<Description>" + itemDescription + "</Description>" +
+			"</UploadVAppTemplateParams>")
+
+	request := client.NewRequest(map[string]string{}, "POST", *createHREF, reqBody)
+	request.Header.Add("Content-Type", "application/vnd.vmware.vcloud.uploadVAppTemplateParams+xml")
+
+	response, err := client.Http.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	resBody, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the XML.
+	catalogItemParsed := &types.CatalogItem{}
+	err = xml.Unmarshal(resBody, &catalogItemParsed)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("[TRACE] Response: %#v \n", response)
+	log.Printf("[TRACE] Response body: %s\n", string(resBody[:]))
+	log.Printf("[TRACE] Catalog item href to query vaapTemplate: %s\n", catalogItemParsed.Entity.HREF)
+
+	ovfUploadUrl, err := url.ParseRequestURI(catalogItemParsed.Entity.HREF)
+	if err != nil {
+		return nil, err
+	}
+
+	return ovfUploadUrl, nil
+}
+
+/** Create Request with right headers and range settings. Support multi part file upload. */
+// TODO Creates a new file upload http request with optional extra params
```

Comment:
Is this really a TODO?

Reply:
Removed comment of todo. Added to my notes.